pkg/execution/state/redis_state/queue_processor.go

```diff
@@ -629,6 +715,152 @@ func (q *queue) processShadowPartition(ctx context.Context, shadowPart *QueueSha
 	return nil
 }
 
+// normalizeBacklog must be called with exclusive access to the shadow partition
+// NOTE: ideally this is one transaction in a lua script but enqueue_to_backlog is way too much work to
+// utilize
+func (q *queue) normalizeBacklog(ctx context.Context, backlog *QueueBacklog) error {
+	rc := q.primaryQueueShard.RedisClient
+
+	l := logger.StdlibLogger(ctx).With("backlog", backlog)
+
+	// TODO: extend the lease
+
+	shard := q.primaryQueueShard
+	var processed int64
+	for {
+		keys := []string{
+			rc.kg.BacklogSet(backlog.BacklogID),
+			rc.kg.QueueItem(),
+		}
+
+		args, err := StrSlice([]any{q.backlogNormalizeLimit})
+		if err != nil {
+			return err
+		}
+
+		byt, err := scripts["queue/backlogNormalizePeek"].Exec(
+			redis_telemetry.WithScriptName(ctx, "backlogNormalizePeek"),
+			rc.Client(),
+			keys,
+			args,
+		).AsBytes()
+		if err != nil {
+			return err
+		}
+
+		type result struct {
+			Total int64                `json:"total"`
+			IDs   []string             `json:"ids"`
+			Items []*osqueue.QueueItem `json:"items"`
+		}
+
+		var res result
+		if err := json.Unmarshal(byt, &res); err != nil {
+			return fmt.Errorf("error unmarshalling backlog normalize peek result: %w", err)
+		}
```

Comment:
tweaked `normalizeBacklog` to retrieve items via a lua script instead so it doesn't necessarily need to hit redis on every single retrieval call.

don't really see a good way around calling enqueueing on each item so leaving that as is. should be fine.